### ISSUE TITLE
Global 'this' capture logic is slightly wrong in `--noImplicitThis`

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#63001]
```ts
// @strict: true
function f(this: typeof globalThis) {
    return () => this;
}
```

**Expected:** no error
**Actual:**

>The containing arrow function captures the global value of 'this'.

I think we just need to add an `isSourceFile(container)` check or something.

**Comment by RyanCavanaugh** [comment]
🤖 Thank you for your issue! I've done some analysis to help get you started. This response is automatically generated; feel free to 👍 or 👎 this comment according to its usefulness.

## Similar Issues

Here are the most similar issues I found
- (69%) [microsoft/typescript#26447](https://redirect.github.com/microsoft/typescript/issues/26447): Wrong error message for \`this\` in function
- (67%) [microsoft/typescript#31719](https://redirect.github.com/microsoft/typescript/issues/31719): Inferring "this" from arrow function is {}
- (67%) [microsoft/typescript#48783](https://redirect.github.com/microsoft/typescript/issues/48783): \`globalThis\` sometimes being stripped from type emition can leads to errors in emited definition

If your issue is a duplicate of one of these, feel free to close this issue. Otherwise, no action is needed.

## Possibly Working as Intended

It looks like this behavior might be intentional. Here's what a computer had to say about it.
